---
title: "countland: R tutorial"
output: rmarkdown::github_document
vignette: >
  %\VignetteIndexEntry{countland: R tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

### Samuel H. Church

This tutorial demonstrates the major functions of `countland` by applying them to a Gold standard single-cell RNA sequencing dataset from [Freytag _et al_ (2018)](10.12688/f1000research.15809.2).

`countland` is a **barebones** set of functions for applying a restricted linear algebra to the analysis of count-based data. As such, there are many opportunities for further optimization that may prove useful in the anlaysis of your own data. We provide the source code freely available at https://github.com/shchurch/countland and encourage users and developers to fork the code for their own purposes.

The following packages are required to complete the tutorial

```{r setup}
library(countland)
library(Seurat)
library(ggplot2)
theme_set(theme_classic())
```

## Get the data

We have used the Gold standard dataset provided by [Freytag _et al_ (2018)](10.12688/f1000research.15809.2). This data consists of ~1000 cells that have ground truth labels corresponding to three human lung cell lines.

`countland` accepts an sparse data matrix (as does `Seurat`). 

```{r data}
gold_path <- system.file("exdata", "Gold_Freytag2018/", package = "countland", mustWork = TRUE)
gold.data <- Seurat::Read10X(data.dir = gold_path)
m <- gold.data
```

## Initialize `countland` object

Initialize `countland` by calling the core function on sparse data matrix.

```{r init}
C <- countland(m)
```
The count matrix is stored in `C@counts`

```{r count-matrix}
C@counts[1:10,1:10]
```

Note that most counts are zero for scRNA-seq data.

## Summarize counts across cells and genes

We can explore our data by checking the total number of counts and other expression measures across cells.

Expression measures include: 
* total counts
* maximum number of counts observed across genes
* number of genes with observed counts
* number of genes with counts above 1, or 10
* number of unique count values per cell
* count index = number of _n_ genes â‰¥_n_ counts

It can be helpful to see how many counts are derived from certain genes, such as mitochondrial genes (here we have chosen a gene name with particularly high counts, regular expressions are permitted for name matching).

```{r cell-score}
# sum counts from genes matching this regex expression
gene_string <- "162396$"

# calculate expression scores across cells
C <- ScoreCells(C,gene_string)
head(C@cell_scores)
```

We can also calculate the same meaures, but across cells.

```{r gene-score}
# calculate expression scores across genes
C <- ScoreGenes(C)
head(C@gene_scores)
```

## Cluster cells by similarity

The dot (or inner) product is a measure of similarity between vectors. In this case, it tells us how similar two cells are based on the distribution of transcript counts, and scaled by the total counts per cell. A dot product of 0 indicates orthogonal cell vectors (no shared counts), larger values indicate aligned cell vectors.

```{r dots}
C <- Dot(C)
```
    
Cell populations can be compared and distinguished by embedding and clustering the matrix of pairwise dot products (contained in `C.dots`). This matrix is an unbounded affinity matrix. It is symmetric, and contains only integer values above 0. Spectral embedding clustering is appropriate for this type of matrix.

First, we embed to investigate the optimal number of clusters for our data.

```{r embed}
C <- Embed(C)
```

The eigengap heuristic is can help decide the optimal number of clusters, but it is only a guideline. According to this heuristic, the optimal number of clusters is _k_ where the difference in eigenvalues $ | e_{k+1} - e_{k} | $ is largest.


```{r eigengap}
PlotEigengap(C)
```

For many datasets, you may want to consider other factors, e.g. choosing a minimum number of clusters, whether or not the eigengap reflects this.

Here we have chosen 3 as the optimal number of clusters.

```{r cluster}
C <- Cluster(C,n_clusters=3)
```

We can now visualize clusters using spectral embedding.

```{r plot-cluster,fig.width=6}
PlotEmbedding(C)
```

## Subsampling data

Cells are not sequenced to standard sequencing depth. This is sometimes a problem for downstream comparisons, but not always.

You can create an alternative count matrix where cells have an equal number of counts using the function `Subsample()`. To subsample to the minimum number of total counts across cells, use `cell_counts='min'`, otherwise the number of counts must be larger than the minimum value.

This matrix is stored in `C@subsample`

```{r subsample-cells}
C <- Subsample(C,cell_counts='min')
```

Similarly, there is often substantial heterogeneity in the magnitude of expression across genes. This may result in highly expressed genes having an outsized impact on results.

You can create an alternate count matrix where gene expression is bounded at a maximum total counts across cells, use `Subsample(gene_counts=[maximum]`. An example maximum value might be equal to 10x the number of cells (columns in the count matrix). 

```{r subsample-genes}
C <- Subsample(C,gene_counts=10*ncol(C@counts)) # doing so will overwrite our previous subsampled matrix 

# to subsample both genes and cells, use gene_counts and cell_counts in the same function. Genes will be subsampled first.

#C <- Subsample(C,gene_counts=10*ncol(C@counts),cell_counts='min')
```

## Subsetting data

You can filter the count matrix to only certain cells and genes using `SubsetCells()` and `SubsetGenes()`.

```{r filt-cell}
filter_cell_names <- C@cell_scores[C@cell_scores$n_features < 8500,]$names
filter_cell_index <- which(C@names_cells %in% filter_cell_names) # cells with fewer than 8,500 unique features
C <- SubsetCells(C,filter_cell_index,remove_empty=FALSE)
```

```{r filt-gene}
filter_gene_names <- C@gene_scores[C@gene_scores$n_cells > 100,]$names
filter_gene_index <- which(C@names_genes %in% filter_gene_names) # cells with greater than 100 observations
C <- SubsetGenes(C,filter_gene_index,remove_empty=FALSE)
```
With `countland`, such data filtering may not be necessary or helpful. The original count matrix can be restored at any time.

```{r restore}
C <- RestoreCounts(C)
```

## Identify marker genes

What makes a gene an ideal marker for a cluster may depend on downstream applications. For example, the ideal marker gene might be defined as the gene detected in all cells in a given cluster and none of the rest. Under this definition, the top marker gene for each cluster can be identified by counting and comparing the number of non-zero cells.

```{r marker-zero}
C <- RankMarkerGenes(C,method='prop-zero',subsample=F)
C@marker_genes[(C@marker_genes$cluster_label == 1),]
```

```{r plot-marker-zero,fig.width=6}
cluster_marker <- C@marker_genes[C@marker_genes$cluster_label == 1,]
cluster_top <- cluster_marker[cluster_marker$rank == 1,]$names
gene_index = which(C@names_genes == cluster_top)
PlotMarker(C,gene_index)
```

Alternatively, the top marker genes for each cluster can be identified by ranking genes according to differential gene expression, calculated using the Wilcoxon rank-sum statistic.

When calculating differential gene expression, it typically makes sense to subsample cells to a standard sequencing depth.

```{r marker-ranks}
C <- Subsample(C,cell_counts='min')
C <- RankMarkerGenes(C,method='rank-sums',subsample=T)
C@marker_genes[(C@marker_genes$cluster_label == 1),]
```

```{r plot-marker-ranks,fig.width=6}
cluster_marker <- C@marker_genes[C@marker_genes$cluster_label == 1,]
cluster_top <- cluster_marker[cluster_marker$rank == 1,]$names
gene_index = which(C@names_genes == cluster_top)
PlotMarker(C,gene_index)
```

